### PR TITLE
Jitclass builtin methods

### PR DESCRIPTION
**Pull request by EPronovost**
Add support for a limited set of python object builtin methods.
We can now do things like
```
@jitclass({"x": intp})
class MyInt:
    def __init__(self, x):
        self.x = x

    def __add__(self, other):
        return MyInt(self.x + other.x)

    ...
```
There are several built-in methods (e.g. `__repr__`, `__call__`) that need additional work and are not covered by this PR.

Fixes https://github.com/numba/numba/issues/1606.
Fixes https://github.com/numba/numba/issues/5690.

(*note*: description edited to link the above issues so they will close upon merge of this PR)

### CODE REVIEW

**Comment by EPronovost** [comment]
The documentation is failing because of `register_simple_overload` (commenting out these fixes it).  I'm not sure how to work around this.  My guess is that it doesn't like the templated functions.

File: numba/experimental/jitclass/overloads.py
Comment (eric-wieser) [inline numba/experimental/jitclass/overloads.py:63]: Is there some way to provide this as a default implementation of `bool` for all objects?

File: numba/experimental/jitclass/overloads.py
Comment (eric-wieser) [inline numba/experimental/jitclass/overloads.py:43]: Is this necessary, vs the simpler:
```suggestion
    dunder = f"__{attr}__"
    def handler(self, *args):
        if not isinstance(self, ClassInstanceType):
            return
        if dunder in jit.methods:
            return lambda self, *args: getattr(self, dunder)(*args)
```

Comment (EPronovost) [reply numba/experimental/jitclass/overloads.py:43]: Unfortunately `getattr` is not currently supported, so that would not work.

File: numba/experimental/jitclass/overloads.py
Comment (EPronovost) [reply numba/experimental/jitclass/overloads.py:63]: I believe that
```
@overload(bool)
def fallback_bool(x):
    return lambda x: True
```
would not be a good idea (if I'm mistaken then cool).  We can add overloads on a type by type basis (e.g. https://github.com/numba/numba/pull/4715)

**Comment by rpopovici** [comment]
Hey @EPronovost, is there a reason why this PR wasn't merged already?

**Comment by gmarkall** [comment]
> Hey @EPronovost, is there a reason why this PR wasn't merged already?

Presently there are test fails in this PR, but the tests ran so long ago it looks like the logs are no longer available. According to @EPronovost's earlier comment it may just be because of docs: https://github.com/numba/numba/pull/5877#issuecomment-650497134

Once the tests are passing, this PR can be queued for review.

**Comment by EPronovost** [comment]
> The documentation is failing because of register_simple_overload (commenting out these fixes it). I'm not sure how to work around this. My guess is that it doesn't like the templated functions. cc: @sklam

@rpopovici as stated above the scripts to generate numba documentation were failing to parse [these lines](https://github.com/numba/numba/pull/5877/files#diff-a659338b412f21c8bb81e8fcbd8c58d78be23c8c80f4346ccff94d2f61e9ff1fR132-R169).  I was hoping a numba core developer would offer a suggestion for ways to resolve this.

**Comment by rpopovici** [comment]
@EPronovost thanks for answering. I was under the impression that this PR is abandonware..
I don't know what numba uses for documentation: sphinx, pdoc? but from the looks of it register_simple_overload has no doc string. Adding one might fix the problem
```
def register_simple_overload(func, *attrs, nargs=1):
"""This is a docstring."""
```
Second, I saw that "eq" and "ne" operators are commented. Is there a reason why they are not supported? or is just a code review mistake

**Comment by EPronovost** [comment]
@rpopovici thanks for the suggestion; docstrings are not the issue here.  Given that someone else is interested in this I looked into it some more and found a workaround.

**Comment by EPronovost** [comment]
Also @stuartarchibald or @sklam I think this PR is ready for review.

**Comment by rpopovici** [comment]
@EPronovost thanks for looking into it. There is one more issue you might one to address. The commented eq and ne comparison operators should be accompanied by an explanation why they are commented in the first place. I am assuming that it has something to do with the missing __hash__ overload..

**Comment by EPronovost** [comment]
@rpopovici Yes previously we had issues trying to overwrite the `hash` method, but I found a way around it.

**Comment by rpopovici** [comment]
Brilliant work @EPronovost! Now there is only one big feature missing from jitclasses. Caching..

**Comment by EPronovost** [comment]
@sklam @stuartarchibald could this PR be marked ready for review?

**Comment by rpopovici** [comment]
@gmarkall @sklam can anyone review this PR? It would be nice to have it in 0.53

**Comment by shaunc** [comment]
I'd bet there aren't more people jumping up and down for this because what comes up first in google are other issues which are dead ends. ... But... 🥳 ... I'm certainly looking forward to it. :)

**Comment by cosmoni** [comment]
reverse methods don't seem to be working with this branch. those should be included before being merged.

**Comment by gmarkall** [comment]
Moving this into 0.55 - it would be nice to give it a push, as it enables the use of lists of jitclasses with `heapq`, as mentioned in this comment: https://github.com/numba/numba/issues/4926#issuecomment-889779403

(cc @truenicoco).

**Comment by gmarkall** [comment]
(Also, the merge conflict is trivial to fix - it's just some additional imports on both sides in `boxing.py` - if you get a minute, @EPronovost , could you merge + resolve conflicts please?)

**Comment by LunarLanding** [comment]
It probably would need its own issue, but would it be possible to have builtin methods work on StructRef's?

**Comment by esc** [comment]
> > I believe the PR is currently stalled
> 
> Is there anything I can do to help? I would really like to see this reach the 0.55 release.

This is an open-source project. You are welcome to do whatever you like (mostly). If you would like to see this in `master` I would suggest trying to contact the original author (they have been away/radio-silent for quite some time) to see what their plans are and also (at the same time) "adopt" the PR. This means, create a new PR with all the commits that have been made so far and then drive it to completion by responding to the reviewer comments until it is merged.

**Comment by gmarkall** [comment]
I'm going to try and adopt this PR - I've just pushed a merge of master to it with resolved conflicts, and will have a look over the reviews shortly. (@EPronovost please let me know if you have any concerns about this)

**Comment by gmarkall** [comment]
As far as I can tell from above, the only outstanding comment was that this PR doesn't presently work for reverse methods. So the path forward here consists of:

- Fix the now-failing tests that result from the merge of master,
- Determine whether support for reverse needs adding, or if it should be omitted from this PR and potentially documented as a limitation,
- A review of the code.

**Comment by gmarkall** [comment]
Oddly, the tests seem to be passing locally for me but all failing in CI.

**Review by gmarkall — DISMISSED** [review]
On the whole I thought this looked good but I've pushed some small changes:

- Fixing up the type error raised in complex typing to be a new style error.
- Some tidy-ups to the test suite.
- Addition of eq / ne testing for the float and int wrappers - these were present in the list of operations to test but were commented out.
- Remove `ipos` and `ineg` from the list of supported dunders, as I don't think these exist.
- List the supported dunders in the docs, to save users having to look at the Numba source code to see the list.

Although I think this is good, it may be worth someone with a little more jitclass experience checking it, e.g. @sklam / @stuartarchibald ?

**Comment by gmarkall** [comment]
> Determine whether support for reverse needs adding, or if it should be omitted from this PR and potentially documented as a limitation,

In the interests of getting this merged, I decided to leave this out of this PR and list the supported dunders - I think it needn't also be explicitly pointed out that this won't be supported.

**Comment by gmarkall** [comment]
Can we try and get this into 0.56? I am aware of active uses of this branch that are waiting on it being merged into Numba.

**Comment by gmarkall** [comment]
Following OOB discussion, @stuartarchibald or @sklam may be able to review this.

File: numba/tests/test_jitclasses.py
Comment (sklam) [inline numba/tests/test_jitclasses.py:1416]: use `numba.core.utils.PYVERSION` instead

File: numba/experimental/jitclass/overloads.py
Comment (sklam) [inline numba/experimental/jitclass/overloads.py:135]: use `numba.core.utils.PYVERSION` instead

File: numba/experimental/jitclass/overloads.py
Comment (sklam) [inline numba/experimental/jitclass/overloads.py:171]: no reflection implemented 

see <https://docs.python.org/3/reference/datamodel.html#object.__ge__>

**Review by sklam — CHANGES_REQUESTED** [review]
Review summary:
- needs to be more testing on the `__bool__` fallback 
- rich-comparison operator lack reflection support. If decided to not support, the limitation should be noted in the docs.

Additional comment:

Too bad the lack of serialization support for jitclass type means we cannot return a Numba typed container containing a jitclass instance.

File: numba/experimental/jitclass/overloads.py
Comment (gmarkall) [reply numba/experimental/jitclass/overloads.py]: This should now be fixed.

**Comment by gmarkall** [comment]
@sklam Summary of changes since your initial review:

- Minor fixups fixed up
- Bool fallback fixed and tested
- Operator reflection implemented and tests
- I also noticed that an implicit `__hash__` member set when `__eq__` is defined but `__hash__` is not caused a problem because it's a class member, which is unsupported. So I also added this case to the ignored attrs, and added a specific test for the issue (this is the last commit on the branch)

I think this should be ready for another look now - many thanks in advance!

**Review by sklam — APPROVED** [review]
Thanks for the fixes.

**Pull request merged**
